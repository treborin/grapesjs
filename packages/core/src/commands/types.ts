/**{START_EVENTS}*/
export enum CommandsEvents {
  /**
   * @event `command:run` Triggered on run of any command.
   * @example
   * editor.on('command:run', ({ id, result, options }) => {
   *  console.log('Command id', id, 'command result', result);
   * });
   */
  run = 'command:run',

  /**
   * @event `command:run:COMMAND-ID` Triggered on run of a specific command.
   * @example
   * editor.on('command:run:my-command', ({ result, options }) => { ... });
   */
  runCommand = 'command:run:',

  /**
   * @event `command:run:before:COMMAND-ID` Triggered before the command is called.
   * @example
   * editor.on('command:run:before:my-command', ({ options }) => { ... });
   */
  runBeforeCommand = 'command:run:before:',

  /**
   * @event `command:abort:COMMAND-ID` Triggered when the command execution is aborted.
   * @example
   * editor.on('command:abort:my-command', ({ options }) => { ... });
   *
   * // The command could be aborted during the before event
   * editor.on('command:run:before:my-command', ({ options }) => {
   *  if (someCondition) {
   *    options.abort = true;
   *  }
   * });
   */
  abort = 'command:abort:',

  /**
   * @event `command:stop` Triggered on stop of any command.
   * @example
   * editor.on('command:stop', ({ id, result, options }) => {
   *  console.log('Command id', id, 'command result', result);
   * });
   */
  stop = 'command:stop',

  /**
   * @event `command:stop:COMMAND-ID` Triggered on stop of a specific command.
   * @example
   * editor.on('command:run:my-command', ({ result, options }) => { ... });
   */
  stopCommand = 'command:stop:',

  /**
   * @event `command:stop:before:COMMAND-ID` Triggered before the command is called to stop.
   * @example
   * editor.on('command:stop:before:my-command', ({ options }) => { ... });
   */
  stopBeforeCommand = 'command:stop:before:',

  /**
   * @event `command:call` Triggered on run or stop of a command.
   * @example
   * editor.on('command:call', ({ id, result, options, type }) => {
   *  console.log('Command id', id, 'command result', result, 'call type', type);
   * });
   */
  call = 'command:call',

  /**
   * @event `command:call:COMMAND-ID` Triggered on run or stop of a specific command.
   * @example
   * editor.on('command:call:my-command', ({ result, options, type }) => { ... });
   */
  callCommand = 'command:call:',
}
/**{END_EVENTS}*/

// need this to avoid the TS documentation generator to break
export default CommandsEvents;
