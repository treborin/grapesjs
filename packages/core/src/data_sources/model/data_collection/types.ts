import { DataCollectionType, DataCollectionVariableType, keyCollectionDefinition } from './constants';
import { ComponentDefinition, ComponentProperties } from '../../../dom_components/model/types';
import { DataVariableProps } from '../DataVariable';

export type DataCollectionDataSource = DataVariableProps | DataCollectionVariableProps;

export interface DataCollectionConfig {
  collectionId: string;
  startIndex?: number;
  endIndex?: number;
  dataSource: DataCollectionDataSource;
}

export enum DataCollectionStateVariableType {
  currentIndex = 'currentIndex',
  startIndex = 'startIndex',
  currentItem = 'currentItem',
  endIndex = 'endIndex',
  collectionId = 'collectionId',
  totalItems = 'totalItems',
  remainingItems = 'remainingItems',
}

export interface DataCollectionState {
  [DataCollectionStateVariableType.currentIndex]: number;
  [DataCollectionStateVariableType.startIndex]: number;
  [DataCollectionStateVariableType.currentItem]: DataVariableProps;
  [DataCollectionStateVariableType.endIndex]: number;
  [DataCollectionStateVariableType.collectionId]: string;
  [DataCollectionStateVariableType.totalItems]: number;
  [DataCollectionStateVariableType.remainingItems]: number;
}

export interface DataCollectionStateMap {
  [key: string]: DataCollectionState;
}

export interface ComponentDataCollectionProps extends ComponentDefinition {
  [keyCollectionDefinition]: DataCollectionProps;
}

export interface ComponentDataCollectionVariableProps
  extends DataCollectionVariableProps,
    Omit<ComponentProperties, 'type'> {}

export interface DataCollectionProps {
  type: typeof DataCollectionType;
  collectionConfig: DataCollectionConfig;
  componentDef: ComponentDefinition;
}

export interface DataCollectionVariableProps {
  type: typeof DataCollectionVariableType;
  variableType: DataCollectionStateVariableType;
  collectionId: string;
  path?: string;
}
