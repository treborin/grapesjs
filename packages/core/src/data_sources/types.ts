import { Model, Collection, ObjectAny } from '../common';
import DataCollectionVariable from './model/data_collection/DataCollectionVariable';
import { DataCollectionVariableProps } from './model/data_collection/types';
import DataRecord from './model/DataRecord';
import DataRecords from './model/DataRecords';
import DataVariable, { DataVariableProps } from './model/DataVariable';
import { DataConditionProps, DataCondition } from './model/conditional_variables/DataCondition';

export type DataResolver = DataVariable | DataCondition | DataCollectionVariable;

export type DataResolverProps = DataVariableProps | DataConditionProps | DataCollectionVariableProps;

export interface DataRecordProps extends ObjectAny {
  /**
   * Record id.
   */
  id: string;

  /**
   * Specifies if the record is mutable. Defaults to `true`.
   */
  mutable?: boolean;

  [key: string]: any;
}

export interface DataSourceListener {
  obj: Model | Collection;
  event: string;
}

interface BaseDataSource {
  /**
   * DataSource id.
   */
  id: string;

  /**
   * DataSource validation and transformation factories.
   */
  transformers?: DataSourceTransformers;

  /**
   * If true will store the data source in the GrapesJS project.json file.
   */
  skipFromStorage?: boolean;
}
export interface DataSourceType<DR extends DataRecordProps> extends BaseDataSource {
  records: DataRecords<DR>;
}
export interface DataSourceProps<DR extends DataRecordProps> extends BaseDataSource {
  records?: DataRecords<DR> | DataRecord<DR>[] | DR[];
}
export type RecordPropsType<T> = T extends DataRecord<infer U> ? U : never;
export interface DataSourceTransformers {
  onRecordSetValue?: (args: { id: string | number; key: string; value: any }) => any;
}

/**{START_EVENTS}*/
export enum DataSourcesEvents {
  /**
   * @event `data:add` Added new data source.
   * @example
   * editor.on('data:add', (dataSource) => { ... });
   */
  add = 'data:add',
  addBefore = 'data:add:before',

  /**
   * @event `data:remove` Data source removed.
   * @example
   * editor.on('data:remove', (dataSource) => { ... });
   */
  remove = 'data:remove',
  removeBefore = 'data:remove:before',

  /**
   * @event `data:update` Data source updated.
   * @example
   * editor.on('data:update', (dataSource, changes) => { ... });
   */
  update = 'data:update',

  /**
   * @event `data:path` Data record path update.
   * @example
   * editor.on('data:path:SOURCE_ID:RECORD_ID:PROP_NAME', ({ dataSource, dataRecord, path }) => { ... });
   */
  path = 'data:path',

  /**
   * @event `data` Catch-all event for all the events mentioned above.
   * @example
   * editor.on('data', ({ event, model, ... }) => { ... });
   */
  all = 'data',
}
/**{END_EVENTS}*/
type DotSeparatedKeys<T> = T extends object
  ? {
      [K in keyof T]: K extends string
        ? T[K] extends object
          ? `${K}` | `${K}.${DotSeparatedKeys<T[K]>}`
          : `${K}`
        : never;
    }[keyof T]
  : never;

export type DeepPartialDot<T> = {
  [P in DotSeparatedKeys<T>]?: P extends `${infer K}.${infer Rest}`
    ? K extends keyof T
      ? Rest extends DotSeparatedKeys<T[K]>
        ? DeepPartialDot<T[K]>[Rest]
        : never
      : never
    : P extends keyof T
      ? T[P]
      : never;
};
