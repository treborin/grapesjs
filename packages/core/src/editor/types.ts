/**{START_EVENTS}*/
export enum EditorEvents {
  /**
   * @event `update` Event triggered on any change of the project (eg. component added/removed, style changes, etc.)
   * @example
   * editor.on('update', () => { ... });
   */
  update = 'update',

  /**
   * @event `undo` Undo executed.
   * @example
   * editor.on('undo', () => { ... });
   */
  undo = 'undo',

  /**
   * @event `redo` Redo executed.
   * @example
   * editor.on('redo', () => { ... });
   */
  redo = 'redo',

  /**
   * @event `load` Editor is loaded. At this stage, the project is loaded in the editor and elements in the canvas are rendered.
   * @example
   * editor.on('load', () => { ... });
   */
  load = 'load',

  /**
   * @event `project:load` Project JSON loaded in the editor. The event is triggered on the initial load and on the `editor.loadProjectData` method.
   * @example
   * editor.on('project:load', ({ project, initial }) => { ... });
   */
  projectLoad = 'project:load',

  /**
   * @event `project:get` Event triggered on request of the project data. This can be used to extend the project with custom data.
   * @example
   * editor.on('project:get', ({ project }) => { project.myCustomKey = 'value' });
   */
  projectGet = 'project:get',

  /**
   * @event `log` Log message triggered.
   * @example
   * editor.on('log', (msg, opts) => { ... });
   */
  log = 'log',

  /**
   * @event `telemetry:init` Initial telemetry data are sent.
   * @example
   * editor.on('telemetry:init', () => { ... });
   */
  telemetryInit = 'telemetry:init',

  /**
   * @event `destroy` Editor started destroy (on `editor.destroy()`).
   * @example
   * editor.on('destroy', () => { ... });
   */
  destroy = 'destroy',

  /**
   * @event `destroyed` Editor destroyed.
   * @example
   * editor.on('destroyed', () => { ... });
   */
  destroyed = 'destroyed',
}
/**{END_EVENTS}*/
