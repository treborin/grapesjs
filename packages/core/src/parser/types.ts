/**{START_EVENTS}*/
export enum ParserEvents {
  /**
   * @event `parse:html` On HTML parse, an object containing the input and the output of the parser is passed as an argument.
   * @example
   * editor.on('parse:html', ({ input, output }) => { ... });
   */
  html = 'parse:html',
  htmlRoot = 'parse:html:root',

  /**
   * @event `parse:html:before` Event triggered before the HTML parsing starts. An object containing the input is passed as an argument.
   * @example
   * editor.on('parse:html:before', (options) => {
   *   console.log('Parser input', options.input);
   *   // You can also process the input and update `options.input`
   *   options.input += '<div>Extra content</div>';
   * });
   */
  htmlBefore = 'parse:html:before',

  /**
   * @event `parse:css` On CSS parse, an object containing the input and the output of the parser is passed as an argument.
   * @example
   * editor.on('parse:css', ({ input, output }) => { ... });
   */
  css = 'parse:css',

  /**
   * @event `parse:css:before` Event triggered before the CSS parsing starts. An object containing the input is passed as an argument.
   * @example
   * editor.on('parse:css:before', (options) => {
   *   console.log('Parser input', options.input);
   *   // You can also process the input and update `options.input`
   *   options.input += '.my-class { color: red; }';
   * });
   */
  cssBefore = 'parse:css:before',

  /**
   * @event `parse` Catch-all event for all the events mentioned above. An object containing all the available data about the triggered event is passed as an argument to the callback.
   * @example
   * editor.on('parse', ({ event, ... }) => { ... });
   */
  all = 'parse',
}
/**{END_EVENTS}*/
