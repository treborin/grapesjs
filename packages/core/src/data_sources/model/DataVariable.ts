import { Model } from '../../common';
import EditorModel from '../../editor/model/Editor';

export const DataVariableType = 'data-variable' as const;

export interface DataVariableProps {
  type: typeof DataVariableType;
  path: string;
  defaultValue?: string;
}

export default class DataVariable extends Model<DataVariableProps> {
  em?: EditorModel;

  defaults() {
    return {
      type: DataVariableType,
      defaultValue: '',
      path: '',
    };
  }

  constructor(props: DataVariableProps, options: { em?: EditorModel }) {
    super(props, options);
    this.em = options.em;
  }

  getDataValue() {
    const { path, defaultValue } = this.attributes;
    return this.em?.DataSources.getValue(path!, defaultValue);
  }
}
