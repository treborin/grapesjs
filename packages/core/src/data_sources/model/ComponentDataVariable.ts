import Component from '../../dom_components/model/Component';
import { ComponentOptions } from '../../dom_components/model/types';
import { toLowerCase } from '../../utils/mixins';
import DataVariable, { DataVariableProps, DataVariableType } from './DataVariable';

export default class ComponentDataVariable extends Component {
  dataResolver: DataVariable;

  get defaults() {
    return {
      // @ts-ignore
      ...super.defaults,
      type: DataVariableType,
      path: '',
      defaultValue: '',
    };
  }

  constructor(props: DataVariableProps, opt: ComponentOptions) {
    super(props, opt);
    const { type, path, defaultValue } = props;
    this.dataResolver = new DataVariable({ type, path, defaultValue }, opt);
  }

  getDataValue() {
    return this.dataResolver.getDataValue();
  }

  getInnerHTML() {
    return this.getDataValue();
  }

  static isComponent(el: HTMLElement) {
    return toLowerCase(el.tagName) === DataVariableType;
  }
}
