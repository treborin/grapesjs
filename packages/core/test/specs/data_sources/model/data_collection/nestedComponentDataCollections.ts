import { Component, DataRecord, DataSource, DataSourceManager, Editor } from '../../../../../src';
import { DataVariableType } from '../../../../../src/data_sources/model/DataVariable';
import {
  DataCollectionType,
  DataCollectionVariableType,
} from '../../../../../src/data_sources/model/data_collection/constants';
import { DataCollectionStateVariableType } from '../../../../../src/data_sources/model/data_collection/types';
import EditorModel from '../../../../../src/editor/model/Editor';
import { setupTestEditor } from '../../../../common';

describe('Collection component', () => {
  let em: EditorModel;
  let editor: Editor;
  let dsm: DataSourceManager;
  let dataSource: DataSource;
  let nestedDataSource: DataSource;
  let wrapper: Component;
  let firstRecord: DataRecord;
  let secondRecord: DataRecord;
  let firstNestedRecord: DataRecord;
  let secondNestedRecord: DataRecord;

  beforeEach(() => {
    ({ em, editor, dsm } = setupTestEditor());
    wrapper = em.getWrapper()!;
    dataSource = dsm.add({
      id: 'my_data_source_id',
      records: [
        { id: 'user1', user: 'user1', age: '12' },
        { id: 'user2', user: 'user2', age: '14' },
      ],
    });

    nestedDataSource = dsm.add({
      id: 'nested_data_source_id',
      records: [
        { id: 'nested_user1', user: 'nested_user1', age: '12' },
        { id: 'nested_user2', user: 'nested_user2', age: '14' },
        { id: 'nested_user3', user: 'nested_user3', age: '16' },
      ],
    });

    firstRecord = dataSource.getRecord('user1')!;
    secondRecord = dataSource.getRecord('user2')!;
    firstNestedRecord = nestedDataSource.getRecord('nested_user1')!;
    secondNestedRecord = nestedDataSource.getRecord('nested_user2')!;
  });

  afterEach(() => {
    em.destroy();
  });

  test('Nested collections bind to correct data sources', () => {
    const parentCollection = wrapper.components({
      type: DataCollectionType,
      collectionDef: {
        componentDef: {
          type: DataCollectionType,
          collectionDef: {
            componentDef: {
              type: 'default',
              name: {
                type: DataCollectionVariableType,
                variableType: DataCollectionStateVariableType.currentItem,
                collectionId: 'nested_collection',
                path: 'user',
              },
            },
            collectionConfig: {
              collectionId: 'nested_collection',
              dataSource: {
                type: DataVariableType,
                path: 'nested_data_source_id',
              },
            },
          },
        },
        collectionConfig: {
          collectionId: 'parent_collection',
          dataSource: {
            type: DataVariableType,
            path: 'my_data_source_id',
          },
        },
      },
    })[0];

    const nestedCollection = parentCollection.components().at(0);
    const nestedFirstChild = nestedCollection.components().at(0);
    const nestedSecondChild = nestedCollection.components().at(1);

    expect(nestedFirstChild.get('name')).toBe('nested_user1');
    expect(nestedSecondChild.get('name')).toBe('nested_user2');
  });

  test('Updates in parent collection propagate to nested collections', () => {
    const parentCollection = wrapper.components({
      type: DataCollectionType,
      collectionDef: {
        componentDef: {
          type: DataCollectionType,
          collectionDef: {
            componentDef: {
              type: 'default',
              name: {
                type: DataCollectionVariableType,
                variableType: DataCollectionStateVariableType.currentItem,
                collectionId: 'nested_collection',
                path: 'user',
              },
            },
            collectionConfig: {
              collectionId: 'nested_collection',
              dataSource: {
                type: DataVariableType,
                path: 'nested_data_source_id',
              },
            },
          },
        },
        collectionConfig: {
          collectionId: 'parent_collection',
          dataSource: {
            type: DataVariableType,
            path: 'my_data_source_id',
          },
        },
      },
    })[0];

    const nestedCollection = parentCollection.components().at(0);
    const nestedFirstChild = nestedCollection.components().at(0);
    const nestedSecondChild = nestedCollection.components().at(1);

    firstNestedRecord.set('user', 'updated_user1');
    expect(nestedFirstChild.get('name')).toBe('updated_user1');
    expect(nestedSecondChild.get('name')).toBe('nested_user2');
  });

  test('Nested collections are correctly serialized', () => {
    const parentCollection = wrapper.components({
      type: DataCollectionType,
      collectionDef: {
        componentDef: {
          type: DataCollectionType,
          collectionDef: {
            componentDef: {
              type: 'default',
              name: {
                type: DataCollectionVariableType,
                variableType: DataCollectionStateVariableType.currentItem,
                path: 'user',
              },
            },
            collectionConfig: {
              collectionId: 'nested_collection',
              dataSource: {
                type: DataVariableType,
                path: 'nested_data_source_id',
              },
            },
          },
        },
        collectionConfig: {
          collectionId: 'parent_collection',
          dataSource: {
            type: DataVariableType,
            path: 'my_data_source_id',
          },
        },
      },
    })[0];

    const serialized = parentCollection.toJSON();
    expect(serialized).toMatchSnapshot();
  });

  test('Nested collections respect startIndex and endIndex', () => {
    const parentCollection = wrapper.components({
      type: DataCollectionType,
      collectionDef: {
        componentDef: {
          type: DataCollectionType,
          collectionDef: {
            componentDef: {
              type: 'default',
              name: {
                type: DataCollectionVariableType,
                variableType: DataCollectionStateVariableType.currentItem,
                collectionId: 'nested_collection',
                path: 'user',
              },
            },
            collectionConfig: {
              collectionId: 'nested_collection',
              startIndex: 0,
              endIndex: 1,
              dataSource: {
                type: DataVariableType,
                path: 'nested_data_source_id',
              },
            },
          },
        },
        collectionConfig: {
          collectionId: 'parent_collection',
          dataSource: {
            type: DataVariableType,
            path: 'my_data_source_id',
          },
        },
      },
    })[0];

    const nestedCollection = parentCollection.components().at(0);
    expect(nestedCollection.components().length).toBe(2);
  });

  test('Nested collection gets and watches value from the parent collection', () => {
    const parentCollection = wrapper.components({
      type: DataCollectionType,
      collectionDef: {
        componentDef: {
          type: DataCollectionType,
          collectionDef: {
            componentDef: {
              type: 'default',
              name: {
                type: DataCollectionVariableType,
                variableType: DataCollectionStateVariableType.currentItem,
                collectionId: 'parent_collection',
                path: 'user',
              },
            },
            collectionConfig: {
              collectionId: 'nested_collection',
              dataSource: {
                type: DataVariableType,
                path: 'nested_data_source_id',
              },
            },
          },
        },
        collectionConfig: {
          collectionId: 'parent_collection',
          dataSource: {
            type: DataVariableType,
            path: 'my_data_source_id',
          },
        },
      },
    })[0];

    const nestedCollection = parentCollection.components().at(0);
    const firstNestedChild = nestedCollection.components().at(0);

    // Verify initial value
    expect(firstNestedChild.get('name')).toBe('user1');

    // Update value in parent collection and verify nested collection updates
    firstRecord.set('user', 'updated_user1');
    expect(firstNestedChild.get('name')).toBe('updated_user1');
  });

  test('Nested collection switches to using its own collection variable', () => {
    const parentCollection = wrapper.components({
      type: DataCollectionType,
      collectionDef: {
        componentDef: {
          type: DataCollectionType,
          collectionDef: {
            componentDef: {
              type: 'default',
              name: {
                type: DataCollectionVariableType,
                variableType: DataCollectionStateVariableType.currentItem,
                path: 'user',
                collectionId: 'parent_collection',
              },
            },
            collectionConfig: {
              collectionId: 'nested_collection',
              dataSource: {
                type: DataVariableType,
                path: 'nested_data_source_id',
              },
            },
          },
        },
        collectionConfig: {
          collectionId: 'parent_collection',
          dataSource: {
            type: DataVariableType,
            path: 'my_data_source_id',
          },
        },
      },
    })[0];

    const nestedCollection = parentCollection.components().at(0);

    const firstChild = nestedCollection.components().at(0);
    // Replace the collection variable with one from the inner collection
    firstChild.set('name', {
      // @ts-ignore
      type: DataCollectionVariableType,
      variableType: DataCollectionStateVariableType.currentItem,
      path: 'user',
      collectionId: 'nested_collection',
    });

    expect(firstChild.get('name')).toBe('nested_user1');
  });

  describe('Nested Collection Component with Parent and Nested Data Sources', () => {
    let parentCollection: Component;
    let nestedCollection: Component;

    beforeEach(() => {
      // Initialize the parent and nested collections
      parentCollection = wrapper.components({
        type: DataCollectionType,
        collectionDef: {
          componentDef: {
            type: DataCollectionType,
            name: {
              type: DataCollectionVariableType,
              variableType: DataCollectionStateVariableType.currentItem,
              collectionId: 'parent_collection',
              path: 'user',
            },
            collectionDef: {
              componentDef: {
                type: 'default',
                name: {
                  type: DataCollectionVariableType,
                  variableType: DataCollectionStateVariableType.currentItem,
                  collectionId: 'nested_collection',
                  path: 'user',
                },
              },
              collectionConfig: {
                collectionId: 'nested_collection',
                dataSource: {
                  type: DataVariableType,
                  path: 'nested_data_source_id',
                },
              },
            },
          },
          collectionConfig: {
            collectionId: 'parent_collection',
            dataSource: {
              type: DataVariableType,
              path: 'my_data_source_id',
            },
          },
        },
      })[0];

      nestedCollection = parentCollection.components().at(0);
    });

    test('Removing a record from the parent data source updates the parent collection correctly', () => {
      // Verify initial state
      expect(parentCollection.components().length).toBe(2); // 2 parent records initially

      // Remove a record from the parent data source
      dataSource.removeRecord('user1');

      // Verify that the parent collection updates correctly
      expect(parentCollection.components().length).toBe(1); // Only 1 parent record remains
      expect(parentCollection.components().at(0).get('name')).toBe('user2'); // Verify updated name

      // Verify that the nested collection is unaffected
      expect(nestedCollection.components().length).toBe(3); // Nested records remain the same
      expect(nestedCollection.components().at(0).get('name')).toBe('nested_user1'); // Verify nested name
    });

    test('Adding a record to the parent data source updates the parent collection correctly', () => {
      // Verify initial state
      expect(parentCollection.components().length).toBe(2); // 2 parent records initially

      // Add a new record to the parent data source
      dataSource.addRecord({ id: 'user3', user: 'user3', age: '16' });

      // Verify that the parent collection updates correctly
      expect(parentCollection.components().length).toBe(3); // 3 parent records now
      expect(parentCollection.components().at(2).get('name')).toBe('user3'); // Verify new name

      // Verify that the nested collection is unaffected
      expect(nestedCollection.components().length).toBe(3); // Nested records remain the same
      expect(nestedCollection.components().at(0).get('name')).toBe('nested_user1'); // Verify nested name
      expect(parentCollection.components().at(2).components().at(0).get('name')).toBe('nested_user1'); // Verify nested name
    });

    test('Removing a record from the nested data source updates the nested collection correctly', () => {
      // Verify initial state
      expect(nestedCollection.components().length).toBe(3); // 3 nested records initially

      // Remove a record from the nested data source
      nestedDataSource.removeRecord('nested_user1');

      // Verify that the nested collection updates correctly
      expect(nestedCollection.components().length).toBe(2); // Only 2 nested records remain
      expect(nestedCollection.components().at(0).get('name')).toBe('nested_user2'); // Verify updated name
      expect(nestedCollection.components().at(1).get('name')).toBe('nested_user3'); // Verify updated name
    });

    test('Adding a record to the nested data source updates the nested collection correctly', () => {
      // Verify initial state
      expect(nestedCollection.components().length).toBe(3); // 3 nested records initially
      expect(nestedCollection.components().at(0).get('name')).toBe('nested_user1'); // Verify initial name
      expect(nestedCollection.components().at(1).get('name')).toBe('nested_user2'); // Verify initial name
      expect(nestedCollection.components().at(2).get('name')).toBe('nested_user3'); // Verify initial name

      // Add a new record to the nested data source
      nestedDataSource.addRecord({ id: 'user4', user: 'nested_user4', age: '18' });

      // Verify that the nested collection updates correctly
      expect(nestedCollection.components().length).toBe(4); // 4 nested records now
      expect(nestedCollection.components().at(3).get('name')).toBe('nested_user4'); // Verify new name

      // Verify existing records are unaffected
      expect(nestedCollection.components().at(0).get('name')).toBe('nested_user1'); // Verify existing name
      expect(nestedCollection.components().at(1).get('name')).toBe('nested_user2'); // Verify existing name
      expect(nestedCollection.components().at(2).get('name')).toBe('nested_user3'); // Verify existing name
    });
  });
});
