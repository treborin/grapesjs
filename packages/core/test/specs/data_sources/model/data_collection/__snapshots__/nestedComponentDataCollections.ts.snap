// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection component Nested collections are correctly serialized 1`] = `
{
  "collectionDef": {
    "collectionConfig": {
      "collectionId": "parent_collection",
      "dataSource": {
        "path": "my_data_source_id",
        "type": "data-variable",
      },
    },
    "componentDef": {
      "collectionDef": {
        "collectionConfig": {
          "collectionId": "nested_collection",
          "dataSource": {
            "path": "nested_data_source_id",
            "type": "data-variable",
          },
        },
        "componentDef": {
          "name": {
            "path": "user",
            "type": "data-collection-variable",
            "variableType": "currentItem",
          },
          "type": "default",
        },
      },
      "type": "data-collection",
    },
  },
  "type": "data-collection",
}
`;
